feat: add gradient accumulation support

irisTa56 opened the pull request:
## Summary / Motivation

Adds gradient accumulation support to the training script.
This allows simulating larger batch sizes without increasing GPU memory usage, which is useful when training large models or when memory is limited.

## What changed

- Added `gradient_accumulation_steps` configuration parameter to `TrainPipelineConfig` (default: 1)
- Updated `update_policy()` to use `accelerator.accumulate()` context manager
- Updated effective batch size calculation in logging and `MetricsTracker`
- Training loop now skips evaluation and checkpointing during gradient accumulation steps (i.e., treats `step` as an optimizer step)
- Added validation that rejects CUDAGraphs compile modes (`max-autotune`, `reduce-overhead`) when `gradient_accumulation_steps > 1`
- Added gradient accumulation section to multi-GPU training docs

## How was this tested (or how to run locally)

Reproduce with a quick example or CLI (if applicable):

```bash
$ lerobot-train \
  --dataset.repo_id=lerobot/svla_so101_pickplace \
  --policy.type=act \
  --policy.repo_id=example/my_policy \
  --policy.push_to_hub=false \
  --batch_size=8 \
  --gradient_accumulation_steps=4 \
  --log_freq=10 \
  --steps=50

# with compilation
$ lerobot-train \
  --dataset.repo_id=lerobot/svla_so101_pickplace \
  --policy.type=smolvla \
  --policy.repo_id=example/my_policy \
  --policy.pretrained_path=lerobot/smolvla_base \
  --policy.compile_model=true \
  --policy.compile_mode=max-autotune-no-cudagraphs \
  --policy.push_to_hub=false \
  --batch_size=8 \
  --gradient_accumulation_steps=4 \
  --log_freq=10 \
  --steps=50
```

## Checklist (required before merge)

- [x] Linting/formatting run (`pre-commit run -a`)
- [ ] All tests pass locally (`pytest`)
  - Some tests were skipped in my local environment: `1586 passed, 123 skipped`
- [x] Documentation updated
- [ ] CI is green

irisTa56 committed:
refactor: simplify update_policy function and metrics handling

irisTa56 committed:
feat: add gradient accumulation support

irisTa56 committed:
Revert some modifications to respect the existing code

irisTa56 committed:
Merge branch 'huggingface:main' into gradient_accumulation

irisTa56 committed:
fix: ensure lr_scheduler steps only when gradients are synced during accumulation

irisTa56 committed:
refactor: improve readability of gradient update logic in update_policy function

irisTa56 committed:
refactor: rename variable for clarity in gradient accumulation test

irisTa56 committed:
revert removing lock parameter to minimize scope changes

Remove lock parameter and related changes that are outside the scope
of gradient accumulation feature. This keeps the branch focused on
its primary topic and reduces unnecessary diff from main.

irisTa56 committed:
Merge branch 'main' into gradient_accumulation

jadechoghari commented:
Hey @irisTa56 that's a very useful feature! I did it locally to train bigger models. Could you maybe resolve the conflicts with main, so we kickoff a review?
Thanks!

irisTa56 committed:
Merge branch 'main' into gradient_accumulation

irisTa56 committed:
Add note about RA-BC weight normalization behavior with gradient accumulation

irisTa56 commented:
@jadechoghari 
Thanks for the feedback! I've resolved the merge conflicts with main and also added code comments for the newly introduced RA-BC feature. The PR should be ready for review now.

jadechoghari commented:
Thanks! @irisTa56, the tests in this PR seem unnecessary since we’re using the accelerate api, and testing this functionality is the responsibility of accelerate itself. Removing the accelerate-related tests would make the PR cleaner and shorter 😄

irisTa56 committed:
Remove accelerate-related tests as they are accelerate's responsibility

jadechoghari commented on tests/utils/test_logging_utils.py:
```diff
@@ -82,6 +83,30 @@ def test_metrics_tracker_step(mock_metrics):
+    assert tracker.episodes == expected_samples / (1000 / 50)
+    assert tracker.epochs == expected_samples / 1000
+
+
```
same here, I don’t think this test is necessary, we’re already testing the metrics stepper itself in other tests, and this ends up indirectly testing accelerate’s behavior rather than our own logic

irisTa56 replied:
Makes sense. Existing tests already cover the metrics stepper logic itself. Removed the extra tests https://github.com/huggingface/lerobot/pull/2646/commits/d365eb8d1ec828adc6c9c75495d585ec8a513de3

irisTa56 committed:
Remove accelerator-related extra test for metrics tracker

irisTa56 committed:
Merge branch 'main' into gradient_accumulation

irisTa56 committed:
Merge branch 'main' into gradient_accumulation

irisTa56 committed:
Merge branch 'main' into gradient_accumulation

irisTa56 committed:
Merge branch 'main' into gradient_accumulation

irisTa56 committed:
Merge branch 'main' into gradient_accumulation

irisTa56 committed:
fix: use while loop to ensure all training steps complete with gradient accumulation

irisTa56 committed:
test: fix MetricsTracker test for gradient_accumulation_steps

Use MockAccelerator to cover gradient_accumulation_steps without relying on Accelerator.

sdfbiasdf commented:
Hi @irisTa56 — quick follow-up to my earlier comment. I refactored the fix into a cleaner, more extensible shape on the same branch. Two main changes:

**1. Per-policy resolution via the config base class.** Instead of a uniform "always fall back to `max-autotune-no-cudagraphs`" rule in `lerobot_train.py`, each policy now declares its own `DEFAULT_COMPILE_MODE` and `SAFE_COMPILE_MODE` as `ClassVar`s, and `PreTrainedConfig.resolve_compile_mode()` is the single source of truth. The train script becomes policy-agnostic and adding a new policy is just two class attributes.

**2. Diffusion gets `"default"` instead of `"max-autotune-no-cudagraphs"`.** This was the bug in my first version: `reduce-overhead` is the documented diffusion default precisely because the U-Net runs small repetitive loops where `max-autotune`-style tuning is overkill. Forcing `max-autotune-no-cudagraphs` when accumulation is on would have run an expensive autotune search against that intent. `"default"` is the lightweight CUDAGraphs-free analogue and preserves the original spirit.

| Policy | DEFAULT | SAFE (with grad_accum > 1) |
|---|---|---|
| pi0 / pi05 / pi0_fast / smolvla | `max-autotune` | `max-autotune-no-cudagraphs` |
| diffusion | `reduce-overhead` | `default` |

Bonus: tests now call `config.resolve_compile_mode()` directly (no helper duplication), parametrized over all 5 policies × both CUDAGraphs modes.

Same offer as before — just wanted to flag the diffusion case specifically.

Refactor commit: https://github.com/sdfbiasdf/lerobot/commit/3b771c23
Docs commit: https://github.com/sdfbiasdf/lerobot/commit/f1e349fd

irisTa56 committed:
Merge branch 'main' into gradient_accumulation

irisTa56 committed:
fix(train): add validation for gradient accumulation steps with CUDAGraphs compatibility

Co-Authored-By: sdfbiasdf <167161769+sdfbiasdf@users.noreply.github.com>

irisTa56 committed:
docs(multi_gpu_training): add gradient accumulation section with usage examples and compatibility notes

Co-Authored-By: sdfbiasdf <167161769+sdfbiasdf@users.noreply.github.com>

irisTa56 committed:
docs(multi_gpu_training): update effective batch size calculation to include gradient accumulation steps

irisTa56 commented:
Good catch, @sdfbiasdf
Thanks for digging into the compile mode interaction and the per-policy resolution!

The automatic per-policy resolution you built is solid work.
Since it touches multiple config files, I think it'd be easier to review as a separate PR.
Would you be open to submitting that as a follow-up?

For this PR, to keep it focused on the core gradient accumulation feature, I added a hard error when CUDAGraphs compile modes are combined with `gradient_accumulation_steps > 1`, along with documentation.

Thanks again!

sdfbiasdf commented:
Sure, happy to open it as a follow-up. Thanks for the credit!

sdfbiasdf commented:
Hi @jadechoghari — any chance of a review here? I have a small follow-up ready to open once this lands.

irisTa56 commented:
Closing this in favor of #3980, which supersedes it and carries the gradient accumulation implementation forward plus the per-policy compile mode resolution follow-up we discussed here.
